securedrop_client/gui/widgets.py

```diff
@@ -913,6 +913,7 @@ def update_sources(self, sources: List[Source]) -> List[str]:
 
             source_widget = self.itemWidget(source_item)
             self.takeItem(self.row(source_item))
+            assert isinstance(source_widget, SourceWidget)
```

Comment:
That is an excellent question. I think we shouldn't raise any `AssertionError`, and if we did get an instance of the supertype, it wouldn't have a `source_uuid` attribute (which is why the type narrowing was necessary in the first place). So I think we would want, indeed, to cause an unmissable error in that case.

Your questiom made me wonder if these assertions could become an obstacle to refactoring. One one hand, they would need to be changed if e.g. `SourceWidget` was renamed, but they would be just fine if we were to subclass `SourceWidget`.
Since the assertion really points to the type where the method being used (`source_uuid` in this case) is defined, I think that's fine. Would you disagree @zekehuntergreen ?

Reply:
Mmm, that's an intersting point @cfm, thanks for raisimg it. If the assertions won't run in production, I'm not sure I can justify using them here. I had completely missed that particularity of assertions.

I need to think about that a little, and look at the alternatives. (My immediate thought about a conditional control flow was that we would be silencing typing errors, but I realize now that we don't have to. E.g. we could raise an exception. And that wouldn't introduce divergent paths liske `assert`s.)

I'll come back to this, please bear with me for a bit! 🐻

Reply:
What I don't like about assertions:

- This use case -type checking- would be an exceptional one. Probably the only case in which we'd want assertions in the production code. Exceptional cases are error-prone.

What I don't mind about assertions:

- Type checking is not a runtime concern. In that sense, adding assertions that are not checked in production is not worse than adding type annotations that are ignored at runtime.
- They're pretty obviously not part of the domain logic _if they're consistently only used for type narrowing_.

What I don't like much about conditional checks:

- They are less obviously related to type checking. Nothing distinguishes a conditional that narrows types form one that is part of the domain logic.
- What to do when the condition is unmet in not obvious, or rather: not automatic. Making sure we have a consistent policy (ignore the error, or raise an exception)  will become one more non-enforced convention to keep track of.

@cfm @zekehuntergreen Both of you will be spending time with this code, just like me: what do you think, what do you prefer, am I missing options or arguments?

Reply:
I wonder if we could add a semgrep rule to only allow `/assert isinstance\(.*\)/`. :bulb:

Reply:
Subsidiary question: what do you(s) think of https://github.com/freedomofpress/securedrop-client/pull/1611/commits/fc3afca587c3fabfdef3932b28599826a4adaf97 ?

Reply:
I like c3afca587c3fabfdef3932b28599826a4adaf97 very much!

Reply:
I like fc3afca587c3fabfdef3932b28599826a4adaf97 very much!